python_modules/libraries/dagster-dbt/dagster_dbt/asset_utils.py

```diff
@@ -789,3 +802,55 @@ def has_self_dependency(dbt_resource_props: Mapping[str, Any]) -> bool:
     has_self_dependency = dagster_metadata.get("has_self_dependency", False)
 
     return has_self_dependency
+
+
+def get_asset_check_key_for_test(
+    manifest: Mapping[str, Any],
+    dagster_dbt_translator: "DagsterDbtTranslator",
+    test_unique_id: str,
+) -> Optional[AssetCheckKey]:
+    if not test_unique_id.startswith("test"):
+        return None
+
+    test_resource_props = manifest["nodes"][test_unique_id]
+    upstream_unique_ids: AbstractSet[str] = set(test_resource_props["depends_on"]["nodes"])
+
+    # If the test is generic, it will have an attached node that we can use.
+    attached_node_unique_id = test_resource_props.get("attached_node")
+
+    # If the test is singular, infer the attached node from the upstream nodes.
+    if len(upstream_unique_ids) == 1:
+        [attached_node_unique_id] = upstream_unique_ids
+
+    # If the test is singular, but has multiple dependencies, infer the attached node from
+    # from the dbt meta.
+    has_asset_key_config = bool(
+        (
+            test_resource_props.get("config", {}).get("meta", {})
+            or test_resource_props.get("meta", {})
+        )
+        .get("dagster", {})
+        .get("asset_key", [])
```

Comment:
We could make this syntax be experimental if we want to remove it in the future. Doesn't seem too hard to keep support for though.